From d8ba38b7e477f3e50c1414506cd36db6e663f122 Mon Sep 17 00:00:00 2001
From: Rich Bono <rbono@redhat.com>
Date: Thu, 20 Nov 2014 21:23:38 -0500
Subject: [scsi] pm8001: fix update_flash

Message-id: <1416518623-7080-18-git-send-email-rbono@redhat.com>
Patchwork-id: 100400
O-Subject: [RHEL7.1 e-stor PATCH V3 17/22] pm8001: fix update_flash
Bugzilla: 1110943
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>

From: Tomas Henzl <thenzl@redhat.com>

The driver checks the return valu, but after he tries to wait_for_completion
which might never happen.  Also the ioctl buffer is freed at the end of the
function, so the first removal is not needed.

Signed-off-by: Tomas Henzl <thenzl@redhat.com>
Acked-by: Suresh Thiagarajan <Suresh.Thiagarajan@pmcs.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 31d05e5b14f362c7ea99d1d9e977771ef434b820)

Signed-off-by: Rich Bono <rbono@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/pm8001/pm8001_ctl.c b/drivers/scsi/pm8001/pm8001_ctl.c
index ade62c8..d3a08ae 100644
--- a/drivers/scsi/pm8001/pm8001_ctl.c
+++ b/drivers/scsi/pm8001/pm8001_ctl.c
@@ -617,11 +617,11 @@ static int pm8001_update_flash(struct pm8001_hba_info *pm8001_ha)
 
   pm8001_ha->nvmd_completion = &completion;
   ret = PM8001_CHIP_DISP->fw_flash_update_req(pm8001_ha, payload);
+  if (ret)
+   break;
   wait_for_completion(&completion);
-  if (ret || (fwControl->retcode > FLASH_UPDATE_IN_PROGRESS)) {
+  if (fwControl->retcode > FLASH_UPDATE_IN_PROGRESS) {
    ret = fwControl->retcode;
-   kfree(ioctlbuffer);
-   ioctlbuffer = NULL;
    break;
   }
  }
-- 
1.7.1